### PR TITLE
Fix for unzipping archives with lot of files

### PR DESCRIPTION
**Pull request by amit-avit**
**Task name**: AzureRmWebAppDeploymentV4

**Description**:
Issue: If any option File Transforms or Variable Substitution is enabled, the zip package is extracted to perform the selected operation when source type is zip. The task uses decompress-zip node module to perform the extraction and for packages with lot of files, the module fails with EMFILE error.
Fix: Instead of using the decompress-zip module for extraction, changed it to consume tools 7zip for windows and unzip for non-windows agents.

**Documentation changes required:** (Y/N) N

**Added unit tests:** (Y/N) N

**Attached related issue:** (Y/N) #4127 

**Checklist**:
- [x] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [x] Checked that applied changes work as expected

**Issue** #4127

### CODE REVIEW

File: Tasks/Common/webdeployment-common-v2/ziputility.ts [hunk 1]
```diff
@@ -2,28 +2,34 @@ import tl = require('azure-pipelines-task-lib/task');
...
-    unzipper.on('extract', function (log) {
-        tl.debug('extracted ' + zipLocation + ' to ' + unzipLocation + ' Successfully');
-        defer.resolve(unzipLocation);
-    });
-    unzipper.extract({
-      path: unzipLocation
-    });
-    return defer.promise;
+
+    var isWin = tl.getPlatform() == tl.Platform.Windows
```
Comment (asranja) [inline Tasks/Common/webdeployment-common-v2/ziputility.ts:15]: use `let` instead of `var` for local variables.

Comment (asranja) [inline Tasks/Common/webdeployment-common-v2/ziputility.ts:15]: nit: do a triple equal comparison `===`

File: Tasks/Common/webdeployment-common-v2/ziputility.ts [hunk 2]
```diff
@@ -2,28 +2,34 @@ import tl = require('azure-pipelines-task-lib/task');
...
+    tl.debug('win: ' + isWin);
+    var unzipRunner: ToolRunner;
+    if (isWin) {
+        tl.debug('Using 7zip tool for extracting');
+        var win7zipLocation = path.join(__dirname, '7zip/7z.exe');
+        unzipRunner = tl.tool(win7zipLocation)
+            .arg([ 'x', `-o${unzipDirLocation}`, zipFileLocation ]);
+    } else {
```
Comment (asranja) [inline Tasks/Common/webdeployment-common-v2/ziputility.ts:23]: start else from new line.

File: Tasks/Common/webdeployment-common-v2/ziputility.ts [hunk 3]
```diff
@@ -2,28 +2,34 @@ import tl = require('azure-pipelines-task-lib/task');
...
+        tl.debug('Using unzip tool for extracting');
+        var unzipToolLocation = tl.which('unzip', true);
+        unzipRunner = tl.tool(unzipToolLocation)
+            .arg([ zipFileLocation, '-d', unzipDirLocation ]);
```
Comment (asranja) [inline Tasks/Common/webdeployment-common-v2/ziputility.ts:27]:
```
unzipRunner = tl.tool(unzipToolLocation)
            .arg([ zipFileLocation, '-d', unzipDirLocation ]);
```
can be moved outside if-else.

Comment (asranja) [reply Tasks/Common/webdeployment-common-v2/ziputility.ts:27]: As part of the PR, we will now be taking a dependency on unzip tool to be present on non-windows agents. Do we know if this tool will always be available ?
The current implementation does not take any external dependency and uses node-module decompress-zip shipped with the task.
Consider a scenario when user was running a pipeline on a private (self-hosted) image. We should ensure that this pipeline continues to work even after new changes and not ask user to install a new dependency (unzip).

**Comment by asranja** [comment]
Fix for #4127

File: Tasks/Common/webdeployment-common-v2/ziputility.ts (hunk 3 above)
Comment (amit-avit) [reply Tasks/Common/webdeployment-common-v2/ziputility.ts:27]: Non-windows agents supported are macOS and linux. Both come with unzip tool out of the box. We are using the same idea in other places as well.

File: Tasks/Common/webdeployment-common-v2/ziputility.ts (hunk 2 above)
Comment (amit-avit) [reply Tasks/Common/webdeployment-common-v2/ziputility.ts]: else in newline is used in c#, this is correct.
btw, removed if-else in new changes.

**Review by asranja — APPROVED** [review]

**Review by AmrutaKawade — APPROVED** [review]

**Pull request merged**